io/zenoh-transport/src/common/pipeline.rs

```diff
@@ -68,269 +62,376 @@ impl fmt::Display for TransportClosed {
+        self.state
+            .fetch_and(!Self::BATCHING_FLAG, Ordering::Relaxed);
     }
+}
 
-    #[inline]
-    fn channel(&self, is_reliable: bool) -> MutexGuard<'_, TransportChannelTx> {
-        if is_reliable {
-            zlock!(self.priority.reliable)
-        } else {
-            zlock!(self.priority.best_effort)
+impl Drop for BatchPool {
+    fn drop(&mut self) {
+        if *self.state.get_mut() & Self::REFILLED_FLAG != 0 {
+            unsafe { self.refill.get_mut().assume_init_drop() }
         }
     }
 }
 
 #[derive(Debug)]
-struct WaitTime {
+struct Deadline {
     wait_time: Duration,
     max_wait_time: Option<Duration>,
+    expiration: Option<Instant>,
 }
 
-impl WaitTime {
+impl Deadline {
     fn new(wait_time: Duration, max_wait_time: Option<Duration>) -> Self {
         Self {
             wait_time,
             max_wait_time,
+            expiration: None,
         }
     }
 
-    fn advance(&mut self, instant: &mut Instant) {
-        match &mut self.max_wait_time {
-            Some(max_wait_time) => {
-                if let Some(new_max_wait_time) = max_wait_time.checked_sub(self.wait_time) {
-                    *instant += self.wait_time;
-                    *max_wait_time = new_max_wait_time;
-                    self.wait_time *= 2;
-                }
-            }
-            None => {
-                *instant += self.wait_time;
-            }
+    fn get_expiration(&mut self) -> Option<Instant> {
+        if self.expiration.is_none() && self.wait_time > Duration::ZERO {
+            self.expiration = Some(Instant::now() + self.wait_time);
         }
+        self.expiration
     }
 
-    fn wait_time(&self) -> Duration {
-        self.wait_time
+    fn renew(&mut self) {
+        if let Some(expiration) = self.get_expiration() {
+            self.expiration = Some(expiration + self.wait_time);
+            if let Some(max) = self.max_wait_time {
+                self.max_wait_time = max.checked_sub(self.wait_time);
+                self.wait_time *= 2;
+            }
+        }
     }
 }
 
-#[derive(Clone)]
-enum DeadlineSetting {
-    Immediate,
-    Finite(Instant),
+struct StageInShared {
+    queue: Mutex<StageIn>,
+    current: Arc<Mutex<Option<WBatch>>>,
+    chan_tx: TransportPriorityTx,
+    batch_pool: Arc<BatchPool>,
 }
 
-struct LazyDeadline {
-    deadline: Option<DeadlineSetting>,
-    wait_time: WaitTime,
+// This is the initial stage of the pipeline where messages are serialized on
+struct StageIn {
+    batch_tx: RingBufferWriter<WBatch, RBLEN>,
+    push_notifier: Notifier,
+    batch_pool: Arc<BatchPool>,
+    refill_waiter: Waiter,
+    use_small_batch: bool,
+    current_notified: bool,
+    fragbuf: ZBuf,
+    batching: bool,
+    batch_config: BatchConfig,
 }
 
-impl LazyDeadline {
-    fn new(wait_time: WaitTime) -> Self {
-        Self {
-            deadline: None,
-            wait_time,
+impl StageIn {
+    const SMALL_BATCH_SIZE: BatchSize = 1 << 11;
```

Comment:
This constant is of course arbitrary. It can be exposed in configuration if needed.